src/parsimmon.js

```diff
@@ -364,38 +364,208 @@ function assertString(x) {
   }
 }
 
+// -*- Error Formatting -*-
+
+var linesBeforeStringError = 2;
+var linesAfterStringError = 3;
+var bytesPerLine = 8;
+var bytesBefore = bytesPerLine * 5;
+var bytesAfter = bytesPerLine * 4;
+var radix = 16;
+var byteLength = 2;
+var defaultLinePrefix = "  ";
+
+function repeat(string, amount) {
+  return new Array(amount + 1).join(string);
+}
+
 function formatExpected(expected) {
   if (expected.length === 1) {
-    return expected[0];
+    return "Expected:\n\n" + expected[0];
+  }
+  return "Expected one of the following: \n\n" + expected.join(", ");
+}
+
+function leftPad(str, pad, char) {
+  var add = pad - str.length;
+  if (add <= 0) {
+    return str;
+  }
+  return repeat(char, add) + str;
+}
+
+function toChunks(arr, chunkSize) {
+  var length = arr.length;
+  var chunks = [];
+  var chunkIndex = 0;
+
+  if (length <= chunkSize) {
+    return [arr.slice()];
+  }
+
+  for (var i = 0; i < length; i++) {
+    if (!chunks[chunkIndex]) {
+      chunks.push([]);
+    }
+
+    chunks[chunkIndex].push(arr[i]);
+
+    if ((i + 1) % chunkSize === 0) {
+      chunkIndex++;
+    }
+  }
+
+  return chunks;
+}
+
+// Get a range of indexes including `i`-th element and `before` and `after` amount of elements from `arr`.
+function rangeFromIndexAndOffsets(i, before, after, length) {
+  return {
+    // Guard against the negative upper bound for lines included in the output.
+    from: i - before > 0 ? i - before : 0,
+    to: i + after > length ? length : i + after
+  };
+}
+
+function byteRangeToRange(byteRange) {
+  // Exception for inputs smaller than `bytesPerLine`
+  if (byteRange.from === 0 && byteRange.to === 1) {
+    return {
+      from: byteRange.from,
+      to: byteRange.to
+    };
   }
-  return "one of " + expected.join(", ");
+
+  return {
+    from: byteRange.from / bytesPerLine,
```

Comment:
Is there a reason this doesn't need to be rounded also?

Reply:
It's because `byteRange.from` should always be dividable by `bytesPerLine`, since it's derived from:

```js
// Removes the reminder from `i`
var byteLineWithErrorIndex = i - (i % bytesPerLine); 

var byteRange = rangeFromIndexAndOffsets(
  byteLineWithErrorIndex, // <- Used to calculate `byteRange.from`
  bytesBefore,
  bytesAfter + bytesPerLine,
  input.length
);

function rangeFromIndexAndOffsets(i, before, after, length) {
  return {
      // `before` is dividable by `bytesPerLine` and `i` is dividable as well
      from: i - before > 0 ? i - before : 0,
      to: i + after > length ? length : i + after
    };
}
```

Maybe this code should be improved, because this particular place is a bit indirect.

Please let me know if you still think there's a problem with this.